File: app/services/auth_service.py
from __future__ import annotations
import os, jwt
from aiokafka import AIOKafkaProducer
from app.config.settings import get_settings, Settings
from app.config.kafka import get_kafka_producer
from app.config.redis import get_redis_client
from app.models.schema import User, ActivationCode, ResetCode, VerificationCode, SecondFactorCode, UserAction, UserDetails, UserDevice, UserRole, UserSettings
from app.config.database import get_session_factory
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.hazmat.primitives.serialization import Encoding, PrivateFormat, PublicFormat, NoEncryption
from datetime import datetime, timedelta, timezone 
from fastapi import Depends
from loguru import logger
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker
from redis.asyncio import Redis
from typing import Any, Dict, Optional, Tuple


class AuthService:
    def __init__(
        self,
        session_factory: async_sessionmaker[AsyncSession],
        settings: Settings, 
        kafka_producer: AIOKafkaProducer, 
        redis_client: Redis, 
    ):

        self.async_session: async_sessionmaker[AsyncSession] = session_factory
        self.kafka_producer: AIOKafkaProducer = kafka_producer
        self.redis_client: Redis = redis_client 
        self.settings: Settings = settings 

        self.private_key_path: str = self.settings.JWT_PRIVATE_KEY_PATH
        self.public_key_path: str = self.settings.JWT_PUBLIC_KEY_PATH
        self.jwt_private_key: Optional[str] = None
        self.jwt_public_key: Optional[str] = None

        if os.environ.get("TEST_MODE") != "True":
            if not self.private_key_path or not self.public_key_path:
                logger.error("JWT keys paths are not defined in settings.")
                raise ValueError("JWT keys paths are required in this application mode")
            
            try:
                with open(self.private_key_path, 'r') as f:
                    self.jwt_private_key = f.read()
                with open(self.public_key_path, 'r') as f:
                    self.jwt_public_key = f.read()
                logger.info("Successfully loaded JWT keys from files.")
            except FileNotFoundError as e:
                logger.error(f"Failed to load JWT keys: Files not found. Make sure secret is mounted properly: {e}")
                raise FileNotFoundError(f"JWT keys not found: {e}")
            except Exception as e:
                logger.error(f"Unexpected error while loading JWT keys: {e}")
                raise Exception(f"Failed to config JWT keys: {e}")

        else: 
            logger.info("Application is running in testing mode, generating new JWT keys.")
            try:
                private_key_obj = ed25519.Ed25519PrivateKey.generate()
                public_key_obj = private_key_obj.public_key()

                self.jwt_private_key = private_key_obj.private_bytes(
                    encoding=Encoding.PEM,
                    format=PrivateFormat.PKCS8,
                    encryption_algorithm=NoEncryption()
                ).decode('utf-8')
                
                self.jwt_public_key = public_key_obj.public_bytes(
                    encoding=Encoding.PEM,
                    format=PublicFormat.SubjectPublicKeyInfo
                ).decode('utf-8')
                logger.success("Successfully generated JWT keys in testing mode.")
            except Exception as e:
                logger.error(f"Error while generating JWT keys: {e}")
                raise Exception(f"Failed to generate JWT keys in test mode: {e}")

        self.jwt_algorithm: str = "EdDSA"
    

    async def generate_jwt_token(self, user_id: int, user_mail: str, expires_delta: Optional[timedelta] = None) -> str:
        if expires_delta:
            expire = datetime.now(timezone.utc) + expires_delta 
        else:
            expire = datetime.now(timezone.utc) + timedelta(hours=24) 

        user_roles: UserRole =  await UserRole.get_user_role_by_user_id(self.async_session, user_id)
        to_encode: Dict[str, Any] = {"sub": str(user_id), "mail": user_mail, "exp": expire, "user_roles": user_roles}
        
        encoded_jwt = jwt.encode(to_encode, self.jwt_private_key, algorithm=self.jwt_algorithm)
        logger.info(f"Generated JWT token for use with ID: {user_id}, which expires at: {expire}")
        return encoded_jwt


    async def register_user(self, mail: str, password: str) -> Optional[Tuple[User, ActivationCode]]:
        logger.info(f"Attempting to register new user with email: {mail}")
        exisiting_user: User = await User.get_user_by_mail(self.async_session, mail, self.redis_client)
        if exisiting_user:
            logger.info("user with provided email already exists")
            return None
        
        registration_result = await User.register(self.async_session, mail, password, self.redis_client)
        if registration_result:
            user, activation_code = registration_result
            logger.success(f"User {mail} registered successfully.")
            return user, activation_code
        else:
            logger.warning(f"Registration failed for email: {mail}. It might already exist.")
            return None


    async def login_user(self, mail: str, password: str) -> Optional[Dict[str, Any]]:
        logger.info(f"Login attempt for user: {mail}")
        user = await User.verify_password(self.async_session, mail, password, self.redis_client)
        
        if not user:
            logger.warning(f"Login failed for {mail}: Invalid credentials.")
            return {"status": "failure", "message": "Invalid credentials."}

        if not user.active:
            logger.warning(f"Login failed for {mail}: Account is not active.")
            return {"status": "failure", "message": "Account is not active."}

        if user.settings and user.settings.second_factor_enabled:
            logger.info(f"Two-factor authentication is enabled for user: {mail}. Generating code.")
            second_factor_code: SecondFactorCode = await user.generate_second_factor_code(self.async_session)
            
            try:
                logger.success(f"Generated 2FA code for user: {mail}")
                return {"status": "mfa_required", "user_id": user.id}
            except Exception as e:
                logger.error(f"Failed to generate 2FA code for user {mail}: {e}")
                return {"status": "failure", "message": "Failed to process 2FA."}

        logger.success(f"Login successful for user: {mail} (2FA not enabled).")
        jwt_token: str = await self.generate_jwt_token(user.id, user.mail)
        return {"status": "success", "token": jwt_token, "user_settings": user.settings}


    async def verify_second_factor(self, second_factor_code: str) -> Dict[str, Any]:
        logger.info(f"Attempting to verify 2FA code: {second_factor_code[:4]}...")
        user: User = await User.get_user_by_second_factor_code(self.async_session, second_factor_code, self.redis_client)

        if not user:
            logger.warning("2FA verification failed: Code is invalid or expired.")
            return None

        user = await user.remove_second_factor(self.async_session, self.redis_client)
        jwt_token: str = await self.generate_jwt_token(user.id, user.mail)
        
        return {"status": "success", "token": jwt_token, "user_settings": user.settings}


    async def activate_account(self, activation_code_str: str) -> Optional[User]:
        logger.info(f"Attempting to activate account with code: {activation_code_str[:8]}...")
        user: User = await User.get_user_by_activation_code(self.async_session, activation_code_str, self.redis_client)
        if not user:
            logger.error("Activation failed: Code not found or invalid.")
            return None

        user = await user.activate_account(self.async_session, activation_code_str) 
        return user


    async def create_verification_code(self, user_id: int) -> Optional[VerificationCode]:
        logger.info(f"Requesting verification code for user ID: {user_id}")
        user: User = await User.get_user_by_id(self.async_session, user_id, self.redis_client)
        if not user:
            logger.error(f"Cannot create verification code: User with ID {user_id} not found.")
            return None
        
        verification_code = await user.create_verification_code(self.async_session, self.redis_client)
        return verification_code


    async def verify_user_account(self, verification_code_str: str) -> Optional[User]:
        logger.info(f"Attempting to verify account with code: {verification_code_str[:8]}...")
        user = await User.get_user_by_verification_code(self.async_session, verification_code_str, self.redis_client)
        if not user:
            logger.error("Verification failed: Code not found or invalid.")
            return None

        verification_code_obj = VerificationCode(code=verification_code_str, user_id=user.id)
        user = await user.verify_account(self.async_session, verification_code_obj, self.redis_client)
        jwt_token: str = await self.generate_jwt_token(user.id, user.mail)
        
        return {"status": "success", "token": jwt_token, "user_settings": user.settings}
        
        
    async def get_user_by_id(self, user_id: int) -> Optional[User]:
        return await User.get_user_by_id(self.async_session, user_id, self.redis_client)


    async def remove_device(self, user_id: int, device_id: str) -> Optional[User]:
        logger.info(f"Attempting to remove device '{device_id}' from user ID {user_id}")
        user = await User.get_user_by_id(self.async_session, user_id, self.redis_client)
        if not user:
            logger.error(f"Cannot remove device: User with ID {user_id} not found.")
            return None

        user = await user.remove_device(self.async_session, device_id, self.redis_client)
        return user


    async def request_password_reset(self, user_id: int):
        pass


    async def confirm_password_reset(self, reset_code: str):
        pass


async def get_auth_service(
    session_factory: async_sessionmaker[AsyncSession] = Depends(get_session_factory),
    settings: Settings = Depends(get_settings),
    kafka_producer: AIOKafkaProducer = Depends(lambda: get_kafka_producer(settings=settings)), 
    redis_client: Redis = Depends(lambda: get_redis_client(settings=settings)) 
) -> AuthService:
    return AuthService(
        session_factory=session_factory,
        settings=settings,
        kafka_producer=kafka_producer,
        redis_client=redis_client
    )

File: app/services/auth_service.go
package services

import (
	"context"
	"crypto"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"
	"time"
	"users-service/app/infrastructure"
	"users-service/app/models/ent"
	requests "users-service/app/models/requests"
	models "users-service/app/models/utils"
	"users-service/app/repositories"
	"users-service/app/utils"

	"github.com/golang-jwt/jwt/v5"
	"github.com/rs/zerolog"
	"golang.org/x/crypto/argon2"
)

type AuthService struct {
	UsersRepository repositories.UsersRepositoryInterface
	EventNotifier   infrastructure.EventNotifierInterface
	Logger          zerolog.Logger
	PublicKey       crypto.PublicKey
	PrivateKey      crypto.PrivateKey
	HashingPepper   []byte
	TokenDuration   time.Duration
	HashPool        chan func()
}

type AuthServiceInterface interface {
	GenerateJWT(ctx context.Context, userID int, deviceID int, userRoles []models.UserRoleInfo) (string, error)
	ValidateJWT(ctx context.Context, tokenString string) (*models.Claims, error)
	HashPassword(password string) (string, error)
	CompareHashes(storedHash, password string) error

	Register(ctx context.Context, req *requests.Register) (*ent.User, error)
	Login(ctx context.Context, req *requests.Login) (*ent.User, string, error)

	ActivateAccount(ctx context.Context, req *requests.Code) (*ent.User, error)
	VerifyAccount(ctx context.Context, req *requests.Code) (*ent.User, string, error)
	VerifySecondFactor(ctx context.Context, req *requests.Code) (*ent.User, string, error)

	RequestPasswordReset(ctx context.Context, req *requests.Mail) error
	CancelPasswordReset(ctx context.Context, req *requests.Mail) error
	ConfirmPasswordReset(ctx context.Context, req *requests.ConfirmPasswordReset) error

	ResendActivationCode(ctx context.Context, req *requests.Mail) error
	ResendVerificationCode(ctx context.Context, req *requests.Mail) error
	ResendResetCode(ctx context.Context, req *requests.Mail) error
	ResendSecondFactorCode(ctx context.Context, req *requests.Mail) error
}

var _ AuthServiceInterface = (*AuthService)(nil)

func NewAuthService(
	usersRepository repositories.UsersRepositoryInterface,
	eventNotifier infrastructure.EventNotifierInterface,
	logger zerolog.Logger,
	publicKey crypto.PublicKey,
	privateKey crypto.PrivateKey,
	poolSize int) *AuthService {

	authService := &AuthService{
		UsersRepository: usersRepository,
		EventNotifier:   eventNotifier,
		Logger:          logger,
		PublicKey:       publicKey,
		PrivateKey:      privateKey,
		HashPool:        make(chan func(), poolSize),
	}

	for i := 0; i < poolSize; i++ {
		go func() {
			for task := range authService.HashPool {
				task()
			}
		}()
	}

	return authService
}

// GenerateJWT creates a signed JWT token for a user with specified roles.
// It sets the user ID, roles, issuance time, and expiration based on TokenDuration.
//
// Parameters:
//   - userID: The user’s ID.
//   - deviceID: The device’s ID.
//   - roles: A slice of role names for the user.
//
// Returns:
//   - A string containing the signed JWT token, or empty if an error occurs.
//   - An error if signing fails, or nil on success.
func (s *AuthService) GenerateJWT(ctx context.Context, userID int, deviceID int, userRoles []models.UserRoleInfo) (string, error) {
	s.Logger.Info().Int("userID", userID).Int("deviceID", deviceID).Msg("Generating JWT for device")

	claims := models.Claims{
		UserID:    userID,
		DeviceID:  deviceID,
		UserRoles: userRoles,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().UTC().Add(s.TokenDuration)),
			IssuedAt:  jwt.NewNumericDate(time.Now().UTC()),
		},
	}

	token := jwt.NewWithClaims(&jwt.SigningMethodEd25519{}, claims)
	resultChan := make(chan models.JWTResult, 1)

	go func() {
		signedToken, err := token.SignedString(s.PrivateKey)
		if err != nil {
			s.Logger.Error().Int("userID", userID).Err(err).Msg("Failed to sign JWT")
			resultChan <- models.JWTResult{Token: "", Err: err}
			return
		}
		s.Logger.Info().Int("userID", userID).Msg("JWT successfully generated")
		resultChan <- models.JWTResult{Token: signedToken, Err: nil}
	}()

	select {
	case result := <-resultChan:
		if result.Err != nil {
			return "", result.Err
		}
		return result.Token, nil
	case <-ctx.Done():
		s.Logger.Warn().Int("userID", userID).Err(ctx.Err()).Msg("JWT generation timed out")
		return "", errors.New("JWT generation timed out")
	}
}

func (s *AuthService) ValidateJWT(ctx context.Context, tokenString string) (*models.Claims, error) {
	s.Logger.Info().Msg("Validating JWT")

	resultChan := make(chan models.JWTValidationResult, 1)

	go func() {
		token, err := jwt.ParseWithClaims(tokenString, &models.Claims{}, func(token *jwt.Token) (interface{}, error) {
			if _, ok := token.Method.(*jwt.SigningMethodEd25519); !ok {
				s.Logger.Warn().Str("alg", fmt.Sprintf("%v", token.Header["alg"])).Msg("Unexpected signing method")
				return nil, errors.New("unexpected signing method")
			}
			return s.PublicKey, nil
		})

		if err != nil {
			s.Logger.Error().Err(err).Msg("JWT parsing failed")
			resultChan <- models.JWTValidationResult{Claims: nil, Err: err}
			return
		}

		claims, ok := token.Claims.(*models.Claims)
		if !ok || !token.Valid {
			s.Logger.Warn().Msg("Invalid token or claims mismatch")
			resultChan <- models.JWTValidationResult{Claims: nil, Err: errors.New("invalid token or claims")}
			return
		}

		if claims.DeviceID == 0 {
			s.Logger.Warn().Int("userID", claims.UserID).Msg("Token is missing device ID")
			resultChan <- models.JWTValidationResult{Claims: nil, Err: errors.New("token is not device-specific")}
			return
		}

		if claims.ExpiresAt != nil && claims.ExpiresAt.Unix() <= time.Now().UTC().Unix() {
			s.Logger.Warn().
				Int("userID", claims.UserID).
				Int("deviceID", claims.DeviceID).
				Time("exp", claims.ExpiresAt.Time).
				Msg("Token expired")
			resultChan <- models.JWTValidationResult{Claims: nil, Err: errors.New("token has expired")}
			return
		}

		s.Logger.Info().
			Int("userID", claims.UserID).
			Int("deviceID", claims.DeviceID).
			Interface("userRoles", claims.UserRoles).
			Msg("Token validated successfully")
		resultChan <- models.JWTValidationResult{Claims: claims, Err: nil}
	}()

	select {
	case result := <-resultChan:
		if result.Err != nil {
			return nil, result.Err
		}
		return result.Claims, nil
	case <-ctx.Done():
		s.Logger.Warn().Err(ctx.Err()).Msg("JWT validation timed out")
		return nil, errors.New("JWT validation timed out")
	}
}

func (s *AuthService) Register(
	ctx context.Context,
	req *requests.Register) (*ent.User, error) {

	if user, _ := s.UsersRepository.GetUserByMail(ctx, req.Mail); user != nil {
		return nil, errors.New("this email is already registered")
	}

	hashedPassword, err := s.HashPassword(req.Password)
	if err != nil {
		return nil, errors.New("failed to hash user's password")
	}

	newUser, activationCode, err := s.UsersRepository.CreateUser(ctx, req, hashedPassword)
	if err != nil {
		return nil, errors.New("failed to create new user")
	}

	if err := s.EventNotifier.CreateRegistrationEvent(newUser, activationCode); err != nil {
		s.Logger.Error().Err(err).Int("userID", newUser.ID).Msg("Failed to produce registration event")
	}

	return newUser, nil
}

func (s *AuthService) ResendActivationCode(ctx context.Context, req *requests.Mail) error {
	user, err := s.UsersRepository.GetUserByMail(ctx, req.Mail)
	if err != nil {
		return errors.New("failed to find user with the provided email")
	}

	if err := s.EventNotifier.CreateRegistrationEvent(user, user.Edges.ActivationCode); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce resend-activation event")
		return errors.New("failed to send activation code")
	}

	s.Logger.Info().Int("userID", user.ID).Msg("Resent activation code event produced")
	return nil
}

func (s *AuthService) ActivateAccount(
	ctx context.Context,
	req *requests.Code) (*ent.User, error) {

	user, err := s.UsersRepository.ActivateAccount(ctx, req.Code)
	if err != nil {
		return nil, errors.New("provided code doesn't exist")
	}

	action := &ent.UserAction{Action: "account.activated"}
	if err := s.EventNotifier.CreateUserActionEvent(user.Edges.UserSettings, action); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce account activation event")
	}

	return user, nil
}

func (s *AuthService) HashPassword(password string) (string, error) {
	resultChan := make(chan struct {
		Hash string
		Err  error
	}, 1)

	s.HashPool <- func() {
		salt := make([]byte, 16)
		if _, err := rand.Read(salt); err != nil {
			s.Logger.Error().Err(err).Msg("Failed to generate salt for password hashing")
			resultChan <- struct {
				Hash string
				Err  error
			}{"", err}
			return
		}

		h := hmac.New(sha256.New, s.HashingPepper)
		h.Write([]byte(password))
		hmacValue := h.Sum(nil)

		hashed := argon2.IDKey(hmacValue, salt, 3, 128*1024, 4, 32)
		hashedStr := fmt.Sprintf("$argon2id$v=19$m=131072,t=3,p=4$%s$%s",
			base64.RawStdEncoding.EncodeToString(salt),
			base64.RawStdEncoding.EncodeToString(hashed))

		resultChan <- struct {
			Hash string
			Err  error
		}{hashedStr, nil}
	}

	result := <-resultChan
	return result.Hash, result.Err
}

func (s *AuthService) CompareHashes(storedHash, password string) error {
	parts := strings.Split(storedHash, "$")
	if len(parts) != 6 {
		return errors.New("invalid stored hash format")
	}

	var memory, hashTime uint32
	var threads uint8
	_, err := fmt.Sscanf(parts[3], "m=%d,t=%d,p=%d", &memory, &hashTime, &threads)
	if err != nil {
		return fmt.Errorf("failed to parse hash parameters: %w", err)
	}

	salt, err := base64.RawStdEncoding.DecodeString(parts[4])
	if err != nil {
		return fmt.Errorf("failed to decode salt from hash: %w", err)
	}

	hash, err := base64.RawStdEncoding.DecodeString(parts[5])
	if err != nil {
		return fmt.Errorf("failed to decode hash part: %w", err)
	}

	resultChan := make(chan error, 1)

	s.HashPool <- func() {
		h := hmac.New(sha256.New, s.HashingPepper)
		h.Write([]byte(password))
		hmacValue := h.Sum(nil)

		comparisonHash := argon2.IDKey(hmacValue, salt, hashTime, memory, threads, uint32(len(hash)))

		if subtle.ConstantTimeCompare(hash, comparisonHash) == 1 {
			resultChan <- nil
		} else {
			resultChan <- errors.New("invalid credentials")
		}
	}

	return <-resultChan
}

func (s *AuthService) Login(ctx context.Context, req *requests.Login) (*ent.User, string, error) {
	user, err := s.UsersRepository.GetUserByMail(ctx, req.Mail)
	if err != nil {
		return nil, "", err
	}

	if user.Blacklisted {
		return nil, "", errors.New("account is blacklisted")
	}

	if !user.Active {
		return nil, "", errors.New("account is not activated")
	}

	if err := s.CompareHashes(user.Password, req.Password); err != nil {
		return nil, "", errors.New("provided wrong password")
	}

	device, err := s.UsersRepository.FindOrCreateDevice(ctx, user.ID, &req.Device)
	if err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to find or create device")
		return nil, "", errors.New("failed to handle user device")
	}

	if user.Edges.UserSettings.TwoFactor {
		userWithCode, secondFactor, err := s.UsersRepository.CreateSecondFactorCode(ctx, user, device)
		if err != nil {
			return nil, "", errors.New("failed to create second factor code")
		}

		if err := s.EventNotifier.CreateSecondFactorEvent(userWithCode.Edges.UserSettings, secondFactor); err != nil {
			s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce second factor event")
			return nil, "", errors.New("failed to send second factor code")
		}

		return nil, "2fa_required", nil
	}

	userRoles := utils.MarshalUserRoles(user.Edges.UserRoles)
	token, err := s.GenerateJWT(ctx, user.ID, device.ID, userRoles)
	if err != nil {
		return nil, "", err
	}

	if err := s.EventNotifier.CreateLoginEvent(user.Edges.UserSettings, "password"); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce login event")
	}

	return user, token, nil
}

func (s *AuthService) ResendSecondFactorCode(ctx context.Context, req *requests.Mail) error {
	user, err := s.UsersRepository.GetUserByMail(ctx, req.Mail)
	if err != nil {
		return errors.New("failed to find user with the provided email")
	}

	if !user.Edges.UserSettings.TwoFactor {
		return errors.New("second factor is not enabled for this account")
	}

	if user.Edges.SecondFactorCode == nil {
		return errors.New("user has no login attempts created")
	}

	if err := s.EventNotifier.CreateSecondFactorEvent(user.Edges.UserSettings, user.Edges.SecondFactorCode); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce resend-second-factor event")
		return errors.New("failed to send second factor code")
	}

	s.Logger.Info().Int("userID", user.ID).Msg("Resent second factor event produced")
	return nil
}

func (s *AuthService) VerifySecondFactor(ctx context.Context, req *requests.Code) (*ent.User, string, error) {
	user, err := s.UsersRepository.GetUserBySecondFactor(ctx, req.Code)
	if err != nil {
		return nil, "", errors.New("failed to fetch user by second factor code")
	}

	if time.Now().UTC().After(user.Edges.SecondFactorCode.ExpiresAt) {
		go s.UsersRepository.RemoveSecondFactorCode(context.Background(), user)
		return nil, "", errors.New("second factor token has expired")
	}

	device := user.Edges.SecondFactorCode.Edges.TargetUserDevice
	if device == nil {
		s.Logger.Error().Int("userID", user.ID).Msg("Device context not found for second factor verification")
		return nil, "", errors.New("could not determine the device for this session")
	}

	userRoles := utils.MarshalUserRoles(user.Edges.UserRoles)
	token, err := s.GenerateJWT(ctx, user.ID, device.ID, userRoles)
	if err != nil {
		return nil, "", errors.New("failed to create Bearer for user")
	}

	user, err = s.UsersRepository.RemoveSecondFactorCode(ctx, user)
	if err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to remove second factor code after use")
	}

	if err := s.EventNotifier.CreateLoginEvent(user.Edges.UserSettings, "second-factor"); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce login event after 2FA")
	}

	return user, token, nil
}

func (s *AuthService) ResendVerificationCode(ctx context.Context, req *requests.Mail) error {
	user, err := s.UsersRepository.GetUserByMail(ctx, req.Mail)
	if err != nil {
		return errors.New("failed to find user with the provided email")
	}

	if err := s.EventNotifier.CreateVerificationEvent(user.Edges.UserSettings, user.Phone, user.Edges.VerificationCode); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce resend-verification event")
		return errors.New("failed to send verification code")
	}

	s.Logger.Info().Int("userID", user.ID).Msg("Resent verification code event produced")
	return nil
}

func (s *AuthService) VerifyAccount(ctx context.Context, req *requests.Code) (*ent.User, string, error) {
	user, err := s.UsersRepository.VerifyAccount(ctx, req.Code)
	if err != nil {
		return nil, "", errors.New("failed to fetch user by verification code")
	}

	device, err := s.UsersRepository.FindOrCreateDevice(ctx, user.ID, &req.Device)
	if err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to find or create device during account verification")
		return nil, "", errors.New("failed to handle user device")
	}

	userRoles := utils.MarshalUserRoles(user.Edges.UserRoles)
	token, err := s.GenerateJWT(ctx, user.ID, device.ID, userRoles)
	if err != nil {
		return nil, "", errors.New("failed to create Bearer for user")
	}

	return user, token, nil
}

func (s *AuthService) RequestPasswordReset(ctx context.Context, req *requests.Mail) error {
	user, err := s.UsersRepository.GetUserByMail(ctx, req.Mail)
	if err != nil {
		return errors.New("failed to fetch user")
	}

	resetCode, err := s.UsersRepository.CreateResetCode(ctx, user)
	if err != nil {
		return errors.New("failed to create reset token")
	}

	if err := s.EventNotifier.CreateResetPasswordEvent(user.Edges.UserSettings, resetCode); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce reset password event")
		return errors.New("failed to send reset password code")
	}

	return nil
}

func (s *AuthService) CancelPasswordReset(ctx context.Context, req *requests.Mail) error {
	user, err := s.UsersRepository.GetUserByMail(ctx, req.Mail)
	if err != nil {
		return errors.New("failed to fetch user")
	}

	if _, err = s.UsersRepository.RemoveResetCode(ctx, user); err != nil {
		return errors.New("failed to cancel password reset process")
	}

	action := &ent.UserAction{Action: "password.reset.cancelled"}
	if err := s.EventNotifier.CreateUserActionEvent(user.Edges.UserSettings, action); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce password reset cancellation event")
	}

	return nil
}

func (s *AuthService) ConfirmPasswordReset(ctx context.Context, req *requests.ConfirmPasswordReset) error {
	user, err := s.UsersRepository.GetUserByResetCode(ctx, req.Code)
	if err != nil {
		return errors.New("failed to fetch user by reset code")
	}

	hashedPassword, err := s.HashPassword(req.Password)
	if err != nil {
		return errors.New("failed to hash user's password")
	}

	if _, err = s.UsersRepository.UpdateUsersPassword(ctx, user, hashedPassword); err != nil {
		return errors.New("failed to update users password")
	}

	if _, err = s.UsersRepository.RemoveResetCode(ctx, user); err != nil {
		return errors.New("failed to remove reset code")
	}

	action := &ent.UserAction{Action: "password.reset.confirmed"}
	if err := s.EventNotifier.CreateUserActionEvent(user.Edges.UserSettings, action); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce password reset confirmation event")
	}

	return nil
}

func (s *AuthService) ResendResetCode(ctx context.Context, req *requests.Mail) error {
	user, err := s.UsersRepository.GetUserByMail(ctx, req.Mail)
	if err != nil {
		return errors.New("failed to find user with the provided email")
	}

	resetCode, err := s.UsersRepository.CreateResetCode(ctx, user)
	if err != nil {
		return errors.New("failed to create a new reset code")
	}

	if err := s.EventNotifier.CreateResetPasswordEvent(user.Edges.UserSettings, resetCode); err != nil {
		s.Logger.Error().Err(err).Int("userID", user.ID).Msg("Failed to produce resend-reset-code event")
		return errors.New("failed to send reset code")
	}

	s.Logger.Info().Int("userID", user.ID).Msg("Resent reset code event produced")
	return nil
}
